pkg/controller/pod.go

```diff
@@ -0,0 +1,264 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	kubeClientSet "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+//Struct for the reconciler
+type ReconcilePod struct {
+	client           client.Client
+	kubeClient       kubeClientSet.Interface
+	azVolumeClient   azClientSet.Interface
+	namespace        string
+	podToClaimsMap   *SharedMap
+	claimToPodsMap   *SharedMap
+	claimToVolumeMap *SharedMap
+	volumeToClaimMap *SharedMap
+}
+
+func (r *ReconcilePod) getPodToClaimsMap() *SharedMap {
+	return r.podToClaimsMap
+}
+func (r *ReconcilePod) getClaimToPodsMap() *SharedMap {
+	return r.claimToPodsMap
+}
+func (r *ReconcilePod) getClaimToVolumeMap() *SharedMap {
+	return r.claimToVolumeMap
+}
+func (r *ReconcilePod) getVolumeToClaimMap() *SharedMap {
+	return r.volumeToClaimMap
+}
+func (r *ReconcilePod) getClient() client.Client {
+	return r.client
+}
+func (r *ReconcilePod) getAzClient() azClientSet.Interface {
+	return r.azVolumeClient
+}
+func (r *ReconcilePod) getNamespace() string {
+	return r.namespace
+}
+
+// Implement reconcile.Reconciler so the controller can reconcile objects
+var _ reconcile.Reconciler = &ReconcilePod{}
+
+func (r *ReconcilePod) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var pod corev1.Pod
+	var value interface{}
+	var exists bool
+
+	podKey := getNamespacedName(request.Namespace, request.Name)
+	if err := r.client.Get(ctx, request.NamespacedName, &pod); err != nil {
+		// if the pod has been deleted, remove entry from podToClaimsMap and claimToPodMap
+		if errors.IsNotFound(err) {
+			value, exists = r.podToClaimsMap.get(podKey)
+			if !exists {
+				return reconcile.Result{}, nil
+			}
+			claims := value.([]string)
+
+			for _, claim := range claims {
+				r.claimToPodsMap.delete(claim)
+			}
+			r.podToClaimsMap.delete(podKey)
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("failed to get Pod (%s): %v", podKey, err)
+		return reconcile.Result{Requeue: true}, err
```

Comment:
Seems like in this PR every time we have an error we specifically set the `Requeue:true` which achieves the same thing as passing a non-nil err. In some of the other controllers for instance the volumeattachment.go we don't follow this convention and just pass a non-nil error. It would be good to have a unified approach in all the controllers. 
Also, we seem to always have a log line when there is an error on reconciling. Would be goof to include line that this request is being requeued.